shared/dataflow/codeql/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -0,0 +1,1912 @@
+          c = decodeContent(token)
+          or
+          not exists(decodeContent(token)) and
+          c = decodeUnknownContent(token)
+        |
+          result = SummaryComponent::content(c)
+        )
+        or
+        exists(ParameterPosition pos |
+          parseArg(token, pos) and
+          result = SummaryComponent::argument(pos)
+        )
+        or
+        exists(ArgumentPosition pos |
+          parseParam(token, pos) and
+          result = SummaryComponent::parameter(pos)
+        )
+        or
+        token = "ReturnValue" and result = SummaryComponent::return(getStandardReturnValueKind())
+        or
+        exists(ReturnKind rk |
+          rk = decodeReturn(token)
+          or
+          not exists(decodeReturn(token)) and
+          rk = decodeUnknownReturn(token)
+        |
+          result = SummaryComponent::return(rk)
+        )
+        or
+        exists(string sg |
+          parseSynthGlobal(token, sg) and result = SummaryComponent::syntheticGlobal(sg)
+        )
+        or
+        exists(ContentSet c |
+          c = decodeWithoutContent(token)
+          or
+          not exists(decodeWithoutContent(token)) and
+          c = decodeUnknownWithoutContent(token)
+        |
+          result = SummaryComponent::withoutContent(c)
+        )
+        or
+        exists(ContentSet c |
+          c = decodeWithContent(token)
+          or
+          not exists(decodeWithContent(token)) and
+          c = decodeUnknownWithContent(token)
+        |
+          result = SummaryComponent::withContent(c)
+        )
+      }
+
+      /**
+       * Holds if `spec` specifies summary component stack `stack`.
+       */
+      predicate interpretSpec(AccessPath spec, SummaryComponentStack stack) {
+        interpretSpec(spec, spec.getNumToken(), stack)
+      }
+
+      /** Holds if the first `n` tokens of `spec` resolves to `stack`. */
+      private predicate interpretSpec(AccessPath spec, int n, SummaryComponentStack stack) {
+        n = 1 and
+        stack = SummaryComponentStack::singleton(interpretComponent(spec.getToken(0)))
+        or
+        exists(SummaryComponent head, SummaryComponentStack tail |
+          interpretSpec(spec, n, head, tail) and
+          stack = SummaryComponentStack::push(head, tail)
+        )
+      }
+
+      /** Holds if the first `n` tokens of `spec` resolves to `head` followed by `tail` */
+      private predicate interpretSpec(
+        AccessPath spec, int n, SummaryComponent head, SummaryComponentStack tail
+      ) {
+        interpretSpec(spec, n - 1, tail) and
+        head = interpretComponent(spec.getToken(n - 1))
+      }
+
+      private class MkStack extends RequiredSummaryComponentStack {
+        override predicate required(SummaryComponent head, SummaryComponentStack tail) {
+          interpretSpec(_, _, head, tail)
+        }
+      }
+
+      private class SummarizedCallableExternal extends SummarizedCallableImpl instanceof SummarizedCallable
```

Comment:
I think this can be simplified. If you consider inlining `summaryElement` then several pieces of this class falls apart. I.e. `relevantSummaryElement` is exactly equal to `super.propagatesFlow`, so that can be deleted and replaced, which also means that `relevantSummaryElementGenerated` can be deleted. The charpred can also be removed, I think, since that's basically represented by the `instanceof SummarizedCallable`.

Reply:
Indeed, this is leftover from the old implementation, thanks for spotting.